### PR TITLE
Include MultiheadAttention module in C# API

### PR DESCRIPTION
**Pull request by fwaris**
We need the MultiheadAttention module by itself for some types of models. 

Although MultiheadAttention is part of the transformer modules, it is also needed separately (i.e. outside of transformer) for some models.

### CODE REVIEW

**Comment by GeorgeS2019** [comment]
It seems TorchSharp[ TorchText is now making critical progress.](https://github.com/xamarin/TorchSharp/issues/257#issue-892759118)

**Comment by GeorgeS2019** [comment]
@fwaris is there a plan to include unit tests and examples for MultiHeadAttention?

**Comment by fwaris** [comment]
I have a unit test that is passing. It exercises the basic functionality.

I am working on porting the [Temporal Graph Network]( https://github.com/twitter-research/tgn) model which requires MHA. Once done, I can create an example also.

File: src/TorchSharp/NN/MultiheadAttention.cs [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+                THSNN_MultiheadAttention_forward(handle,
+                    query.Handle,
+                    key.Handle,
+                    value.Handle,
+                    key_padding_mask?.Handle ?? IntPtr.Zero,
+                    need_weights,
+                    attn_mask?.Handle ?? IntPtr.Zero,
+                    out var res1,
+                    out var res2);
+                if (res1 == IntPtr.Zero) { torch.CheckForErrors(); }
```
Comment (NiklasGustafsson) [inline src/TorchSharp/NN/MultiheadAttention.cs:44]: I think it should probably check booth res1 and res2, just to be safe.

File: src/TorchSharp/NN/MultiheadAttention.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+            /// <param name="embeded_dim">total dimension of the model</param>
+            /// <param name="num_heads">parallel attention heads</param>
+            /// <param name="dropout">a Dropout layer on attn_output_weights. Default: 0.0</param>
+            /// <param name="bias">add bias as module parameter. Default: true</param>
+            /// <param name="add_bias_kv">add bias to the key and value sequences at dim=0</param>
+            /// <param name="add_zero_attn">add a new batch of zeros to the key and value sequences at dim=1</param>
+            /// <param name="kdim">total number of features in key</param>
+            /// <param name="vdim">total number of features in value</param>
+            /// <returns></returns>
+            static public MultiheadAttention MultiheadAttention(long embeded_dim, long num_heads, double dropout = 0.0, bool bias = true, bool add_bias_kv = false, bool add_zero_attn = false, long? kdim=null, long? vdim=null)
```
Comment (NiklasGustafsson) [inline src/TorchSharp/NN/MultiheadAttention.cs:69]: 'embeded_dim' -> 'embedded_dim'

Comment (fwaris) [reply src/TorchSharp/NN/MultiheadAttention.cs:69]: will fix. thanks

**Comment by NiklasGustafsson** [comment]
@fwaris:

This is a good addition to TorchSharp, and we really value your contribution. 

We're in the middle of moving this repository to a different organization (Xamarin is not the right long-term place for it), where we will have a proper Contributor License Agreement for all contributions, internal and external.

Therefore, we will hold off on merging this PR for a little bit.

**Comment by NiklasGustafsson** [comment]
An update: this repo will be moved to the .NET Foundation organization, after which we will have the CLA and can accept the contribution.

**Comment by NiklasGustafsson** [comment]
@fwaris, the easiest thing is for me to close this PR and you to open a new one. We really appreciate your contribution, and I apologize if this is inconvenient for you.

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=320) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=320) before we can accept your contribution.<br/><br/>:x: fwaris [sign now](https://cla.dotnetfoundation.org/dotnet/TorchSharp?pullRequest=320)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/TorchSharp?pullRequest=320) it.</sub>

File: src/TorchSharp/NN/MultiheadAttention.cs [hunk 3]
```diff
@@ -0,0 +1,79 @@
+// Copyright (c) Microsoft Corporation and contributors.  All Rights Reserved.  See License.txt in the project root for license information.
```
Comment (NiklasGustafsson) [inline src/TorchSharp/NN/MultiheadAttention.cs:1]: This should now say 'Copyright (c) .NET Foundation and Contributors. All Rights Reserved. See LICENSE in the project root for license information.'

File: src/TorchSharp/NN/MultiheadAttention.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+            /// <param name="value"></param>
+            /// <param name="key_padding_mask">if provided, specified padding elements in the key will be ignored by the attention. When given a binary mask and a value is True, the corresponding value on the attention layer will be ignored. When given a byte mask and a value is non-zero, the corresponding value on the attention layer will be ignored</param>
+            /// <param name="need_weights">output attn_output_weights</param>
+            /// <param name="attn_mask">2D or 3D mask that prevents attention to certain positions. A 2D mask will be broadcasted for all the batches while a 3D mask allows to specify a different mask for the entries of each batch</param>
+            /// <returns>attn_output, attn_ouput_weights</returns>
+
+            public Tuple<Tensor,Tensor> forward(Tensor query, Tensor key, Tensor value, Tensor? key_padding_mask = null, bool need_weights = true, Tensor? attn_mask = null)
+            //const NNModule module, const Tensor query, const Tensor key, const Tensor value, const Tensor key_padding_mask, const bool need_weights, const Tensor attn_mask, Tensor res1, Tensor res2 )
```
Comment (NiklasGustafsson) [inline src/TorchSharp/NN/MultiheadAttention.cs:31]: There's commented code,

File: src/TorchSharp/NN/MultiheadAttention.cs [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            {
+                //var res1 = IntPtr.Zero;
```
Comment (NiklasGustafsson) [inline src/TorchSharp/NN/MultiheadAttention.cs:33]: More commented code.

File: .gitignore [hunk 6]
```diff
@@ -268,3 +268,4 @@ packages/
 .ionide
 *.bin
 /*.png
+/src/Native/out/build/x64-Debug
```
Comment (NiklasGustafsson) [inline .gitignore:271]: Odd. Was this necessary?

**Review by NiklasGustafsson — CHANGES_REQUESTED** [review]
Minor changes required before the PR will be accepted.

**Comment by fwaris** [comment]
i signed the CLA but not sure if the PR is still valid. If not,  create a new one later this week.

**Comment by NiklasGustafsson** [comment]
> i signed the CLA but not sure if the PR is still valid. If not, create a new one later this week.

I closed and resubmitted the PR, so now the checks are running. I had some review comments.

**Comment by NiklasGustafsson** [comment]
@fwaris, I would like to get this in and then do a NuGet release with it, so I'm going to merge the PR and address the requested changes in another PR.

**Comment by fwaris** [comment]
thanks. I will rebase and make the request changes as another PR

**Comment by NiklasGustafsson** [comment]
> thanks. I will rebase and make the request changes as another PR

@fwaris, there's no need. I already made the changes and they are in main now.

**Comment by GeorgeS2019** [comment]
@fwaris  We are attempting [to visualize MultiheadAttention  module](https://github.com/fel88/Dagre.NET/issues/2) within the transformer architecture. You input is appreciated.

:-)  I still looking forwards to your tests and samples on MultiheadAttention :-)

**Comment by fwaris** [comment]
day job is keeping me very busy but I intend to complete the TGN model port (see above) port in the next few weeks. Only a few modules left to port.

**Comment by GeorgeS2019** [comment]
@fwaris take your time. Thx for valuable voluntary contribution :-)

**Pull request merged**